java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java

```diff
@@ -522,15 +530,44 @@ public void reallocDataBuffer() {
    * @throws OutOfMemoryException if the internal memory allocation fails
    */
   public void reallocValidityAndOffsetBuffers() {
-    offsetBuffer = reallocBufferHelper(offsetBuffer, true);
-    validityBuffer = reallocBufferHelper(validityBuffer, false);
+    int offsetBaseSize = Integer.max(offsetBuffer.capacity(), offsetAllocationSizeInBytes);
+    long newOffsetBufferSize = offsetBaseSize * 2L;
+    long targetOffsetBufferSize = roundUp8(BaseAllocator.nextPowerOfTwo(newOffsetBufferSize));
+    long targetOffsetCount = targetOffsetBufferSize / 4;
+    targetOffsetBufferSize -= getValidityBufferSizeFromCount((int) targetOffsetCount);
+    if (newOffsetBufferSize < targetOffsetBufferSize) {
+      newOffsetBufferSize = targetOffsetBufferSize;
+    }
+
+    long newValidityBufferSize = getValidityBufferSizeFromCount((int)(newOffsetBufferSize / 4));
+
+    long newAllocationSize = newOffsetBufferSize + newValidityBufferSize;
+    assert newAllocationSize >= 1;
+
+    if (newAllocationSize > MAX_ALLOCATION_SIZE) {
+      throw new OversizedAllocationException("Unable to expand the buffer");
+    }
+
+    ArrowBuf[] buffers = allocAndSlice(allocator,
+        new int[] {(int)newOffsetBufferSize, (int)newValidityBufferSize});
+    final ArrowBuf newOffsetBuffer = buffers[0];
+    newOffsetBuffer.setBytes(0, offsetBuffer, 0, offsetBuffer.capacity());
+    newOffsetBuffer.setZero(offsetBuffer.capacity(), (int)newOffsetBufferSize - offsetBuffer.capacity());
+    offsetBuffer.release();
+    offsetBuffer = newOffsetBuffer;
+    offsetAllocationSizeInBytes = (int)newOffsetBufferSize;
+
+    final ArrowBuf newValidityBuffer = buffers[1];
+    newValidityBuffer.setBytes(0, validityBuffer, 0, validityBuffer.capacity());
+    newValidityBuffer.setZero(validityBuffer.capacity(), (int)newValidityBufferSize - validityBuffer.capacity());
+    validityBuffer.release();
+    validityBuffer = newValidityBuffer;
+    validityAllocationSizeInBytes = (int)newValidityBufferSize;
   }
 
-  /* helper method to realloc a particular buffer. returns the allocated buffer */
-  private ArrowBuf reallocBufferHelper(ArrowBuf buffer, final boolean offsetBuffer) {
-    final int currentBufferCapacity = buffer.capacity();
-    long baseSize = (offsetBuffer ? offsetAllocationSizeInBytes
-            : validityAllocationSizeInBytes);
+  private void reallocValidityBuffer() {
+    final int currentBufferCapacity = validityBuffer.capacity();
```

Comment:
Why do we have this method at all? Shouldn't this be shared with the value buffer?

Reply:
This is used by an api called setNull(int index) -  it just sets the validity bit to false without actually adding a data element. If this optimization isn't useful, I can switch to expand both validity/offset arrays (I preserved the old behavior).

Also, used by another api setIndexDefined(int index). I think this one isn't meaningful unless we expand both the validity and the offset arrays.

Reply:
Yes, we should always only resize together. If you resized the validity buffer, you'll need to resize the data buffer ultimately anyway (when you call setValueCount). We shouldn't be ever doing a partial resize.